File: python3/leetcodepy/spiral_matrix_ii.py
"""
Given a positive integer n, generate an n x n matrix filled with elements from 1 to n2 in spiral order.



Example 1:
https://assets.leetcode.com/uploads/2020/11/13/spiraln.jpg

Input: n = 3
Output: [[1,2,3],[8,9,4],[7,6,5]]

Example 2:

Input: n = 1
Output: [[1]]


Constraints:

1 <= n <= 20
"""
from typing import List


class Solution1:
    def generateMatrix(self, n: int) -> List[List[int]]:
        result = [[0] * n for _ in range(n)]
        left, right, top, bottom = 0, n - 1, 0, n - 1
        count = 1
        while True:
            for j in range(left, right + 1):
                result[top][j] = count
                count += 1
            top += 1
            if top > bottom:
                break
            for i in range(top, bottom + 1):
                result[i][right] = count
                count += 1
            right -= 1
            if left > right:
                break
            for j in range(right, left - 1, -1):
                result[bottom][j] = count
                count += 1
            bottom -= 1
            if top > bottom:
                break
            for i in range(bottom, top - 1, -1):
                result[i][left] = count
                count += 1
            left += 1
            if left > right:
                break
        return result

File: java/src/main/java/io/binac/leetcode/SpiralMatrixII.java
package io.binac.leetcode;

/**
 * <p>Given a positive integer <code>n</code>, generate an <code>n x n</code> <code>matrix</code> filled with elements from <code>1</code> to <code>n<sup>2</sup></code> in spiral order.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 * <img alt="" src="https://assets.leetcode.com/uploads/2020/11/13/spiraln.jpg" style="width: 242px; height: 242px;">
 * <pre><strong>Input:</strong> n = 3
 * <strong>Output:</strong> [[1,2,3],[8,9,4],[7,6,5]]
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> n = 1
 * <strong>Output:</strong> [[1]]
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= n &lt;= 20</code></li>
 * </ul>
 */
public class SpiralMatrixII {
    public static class Solution1 {
        public int[][] generateMatrix(int n) {
            int[][] result = new int[n][n];
            int value = 1;
            int left = 0, right = n - 1, top = 0, bottom = n - 1;
            int i, j;
            while (true) {
                for (j = left; j <= right; ++j)
                    result[top][j] = value++;
                ++top;
                if (top > bottom) break;
                for (i = top; i <= bottom; ++i)
                    result[i][right] = value++;
                --right;
                if (left > right) break;
                for (j = right; j >= left; --j)
                    result[bottom][j] = value++;
                --bottom;
                if (top > bottom) break;
                for (i = bottom; i >= top; --i)
                    result[i][left] = value++;
                ++left;
                if (left > right) break;
            }
            return result;
        }
    }
}
